### ISSUE TITLE
`dill.dump(..., recurse=True)` fail when import inside `__setstate__`

### ISSUE THREAD

**Issue by Conchylicultor** [uqfoundation/dill#403]
Reproduction instruction:

```python
import dill


class A:
    def __getstate__(self):
        return 123

    def __setstate__(self, state):
        import os  # Trigger ImportError: __import__ not found in dill.load



with open("tmp_fn.pkl", "wb") as f:
    dill.dump(A(), f, recurse=True)


with open("tmp_fn.pkl", "rb") as f:
    print(dill.load(f))  # ImportError: __import__ not found
```

This fail with:

```
Traceback (most recent call last):
  File "/Users/epot/__A__/Code/experimental/dill_test.py", line 17, in <module>
    print(dill.load(f))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/dill/_dill.py", line 278, in load
    return Unpickler(file, ignore=ignore, **kwds).load()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/dill/_dill.py", line 481, in load
    obj = StockUnpickler.load(self)
  File "/Users/epot/__A__/Code/experimental/dill_test.py", line 9, in __setstate__
    import os
ImportError: __import__ not found
```

**Comment by Bankde** [comment]
I have the similar issue:
Test on both 0.3.3 and 0.3.4.dev0 (current master branch)

```
import dill as pickle

if pickle.__name__ == "dill":
    pickle.settings['recurse'] = True

def test2(b):
    return b+10

def test1(a):
    import numpy as np
    return np.log(test2(a))

funcData = pickle.dumps(test1)
with open("data.txt", "wb+") as f:
    f.write(funcData)

with open("data.txt", "rb") as f:
    data = f.read()

newFunc = pickle.loads(data)
print(newFunc(2.718 - 10))
```

```
Traceback (most recent call last):
  File "/Users/bankde/Desktop/project/pickle/p1.py", line 24, in <module>
    print(newFunc(2.718 - 10))
  File "/Users/bankde/Desktop/project/pickle/p1.py", line 10, in test1
    import numpy as np
ImportError: __import__ not found
```

In my case, I manually fixed by moving import to outside the function.
I'm not sure if this is a recommended solution for dill? Importing inside the function works fine for cloudpickle.

**Comment by mmckerns** [comment]
This should be fixed by the patch added for #219 
```
>>> import dill
>>> 
>>> 
>>> class A:
...     def __getstate__(self):
...         return 123
...     def __setstate__(self, state):
...         import os  # Trigger ImportError: __import__ not found in dill.load
... 
>>> 
>>> dill.settings['recurse'] = True
>>> 
>>> dill.loads(dill.dumps(A()))
<__main__.A object at 0x102651d00>
```
Reopen if this is still an issue for you.
